server/src/main/java/org/apache/druid/server/QueryScheduler.java

```diff
@@ -0,0 +1,226 @@
+  {
+    List<Bulkhead> bulkheads = acquireLanes(query);
+    return resultSequence.withBaggage(() -> releaseLanes(bulkheads));
+  }
+
+  /**
+   * Forcibly cancel all futures that have been registered to a specific query id
+   */
+  public boolean cancelQuery(String id)
+  {
+    // if you re-use queryId and cancel queries... you are going to have a bad time
+    queryDatasources.removeAll(id);
+    Set<ListenableFuture<?>> futures = queryFutures.removeAll(id);
+    boolean success = true;
+    for (ListenableFuture<?> future : futures) {
+      success = success && future.cancel(true);
+    }
+    return success;
+  }
+
+  /**
+   * Get a {@link Set} of datasource names for a {@link Query} id, used by {@link QueryResource#cancelQuery} to
+   * authorize that a user may call {@link #cancelQuery} for the given id and datasources
+   */
+  public Set<String> getQueryDatasources(final String queryId)
+  {
+    return queryDatasources.get(queryId);
+  }
+
+  /**
+   * Get the maximum number of concurrent queries that {@link #run} can support
+   */
+  public int getTotalAvailableCapacity()
+  {
+    return laneRegistry.getConfiguration(TOTAL)
+                       .map(config -> laneRegistry.bulkhead(TOTAL, config).getMetrics().getAvailableConcurrentCalls())
+                       .orElse(-1);
+  }
+
+  /**
+   * Get the maximum number of concurrent queries that {@link #run} can support for a given lane
+   */
+  public int getLaneAvailableCapacity(String lane)
+  {
+    return laneRegistry.getConfiguration(lane)
+                       .map(config -> laneRegistry.bulkhead(lane, config).getMetrics().getAvailableConcurrentCalls())
+                       .orElse(-1);
+  }
+
+  /**
+   * Acquire a semaphore for both the 'total' and a lane, if any is associated with a query
+   */
+  @VisibleForTesting
+  List<Bulkhead> acquireLanes(Query<?> query)
+  {
+    final String lane = QueryContexts.getLane(query);
+    final Optional<BulkheadConfig> laneConfig = lane == null ? Optional.empty() : laneRegistry.getConfiguration(lane);
+    List<Bulkhead> hallPasses = new ArrayList<>(2);
+    final Optional<BulkheadConfig> totalConfig = laneRegistry.getConfiguration(TOTAL);
+    // if we have a lane, get it first
+    laneConfig.ifPresent(config -> {
+      Bulkhead laneLimiter = laneRegistry.bulkhead(lane, config);
+      if (!laneLimiter.tryAcquirePermission()) {
+        throw new QueryCapacityExceededException(lane);
+      }
+      hallPasses.add(laneLimiter);
+    });
+
+    // everyone needs to take one from the total lane; to ensure we don't acquire a lane and never release it, we want
+    // to check for total capacity exceeded and release the lane (if present) before throwing capacity exceeded
+    totalConfig.ifPresent(config -> {
+      Bulkhead totalLimiter = laneRegistry.bulkhead(TOTAL, config);
+      if (!totalLimiter.tryAcquirePermission()) {
+        releaseLanes(hallPasses);
```

Comment:
May be better to do this in a exception handler in case there's an unexpected error between the two acquisitions. I don't think there's a test for this scenario?

Reply:
Hmm, I couldn't find an exception that would be thrown in this block, `tryAcquirePermission` eats `InterruptedException` and also the timeout is configured to 0 so the underlying semaphore call takes a code path that shouldn't explode, so I imagined any exception that happens where would be indicative of a catastrophic failure or the jetty server shutting down. 

That said, I guess it doesn't harm much to wrap in a try catch and rethrow after releasing.

This isn't explicitly tested for, but a shadow of it being tested is visible in the concurrency tests which flex the case where a lane was acquired but the total could not be in [this assert that the 'lane released' count is equal to the 'lane acquired' + 'lane not acquired' count](https://github.com/apache/druid/pull/9407/files#diff-3b8f9c560c324b6f177ec84f50c3c856R551) which fails if we don't consider the 'lane not acquired' count.